src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const clones = [ { ...state } ];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        const copy = { ...clones[clones.length - 1] };
+
+        clones.push(Object.assign(copy, action.extraData));
+        break;
+
+      case 'removeProperties':
+        const lastObj = { ...clones[clones.length - 1] };
+
+        for (const key of action.keysToRemove) {
+          delete lastObj[key];
+        }
+        clones.push(lastObj);
+        break;
+
+      case 'clear':
+        clones.push({});
+        break;
+    }
+  }
+
+  clones.shift();
```

Comment:
Just create a one more variable for the initial state and you don't need `shift()`